pandas/core/indexes/base.py

```diff
@@ -2962,19 +2961,13 @@ def union(self, other, sort=None):
                     stacklevel=2,
                 )
 
-            dtype = find_common_type([self.dtype, other.dtype])
-            if self._is_numeric_dtype and other._is_numeric_dtype:
-                # Right now, we treat union(int, float) a bit special.
-                # See https://github.com/pandas-dev/pandas/issues/26778 for discussion
-                # We may change union(int, float) to go to object.
-                # float | [u]int -> float  (the special case)
-                # <T>   | <T>    -> T
-                # <T>   | <U>    -> object
-                if not (is_integer_dtype(self.dtype) and is_integer_dtype(other.dtype)):
-                    dtype = np.dtype("float64")
-                else:
-                    # one is int64 other is uint64
-                    dtype = np.dtype("object")
+            dtype = find_common_type([self.dtype, other.dtype], strict_uint64=True)
+            # Right now, we treat union(float, [u]int) a bit special.
+            # See https://github.com/pandas-dev/pandas/issues/26778 for discussion
+            # Now it's:
+            # * float | [u]int -> float
+            # * uint64 | signed int  -> object
+            # We may change union(float | [u]int) to go to object.
```

Comment:
i agree this logic doesn't belong here, but suggest putting it in Index._find_common_type_compat instead of higher up, to keep the Index casting logic in one place